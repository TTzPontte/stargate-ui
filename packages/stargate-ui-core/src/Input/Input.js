import React, { useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { useStyles } from '@pontte/stargate-ui-styles';
import { Close as SvgIconClose } from '@pontte/stargate-ui-icons';

import Factory from '../Factory';
import InputLabel from '../InputLabel';
import InputHelper from '../InputHelper';

/**
 * @todo check color pattern for input element
 * @todo needs pattern for disabled and readonly state
 */
const styles = (theme) => {
  const {
    palette,
    radius,
    active,
    mode,
  } = theme;
  const { setLightness } = palette;

  const getColor = (color, type = 'color') => (
    palette?.[color]?.[mode][type]
  );

  const getColorLight = (n, color) => (
    setLightness(n, getColor(color))
  );

  const inlineFlexCenter = {
    display: 'inline-flex',
    alignItems: 'center',
  };

  const inputGroup = {
    ...inlineFlexCenter,
    overflow: 'hidden',
    borderRadius: radius(),
    width: '100%',
    border: [[1, 'solid']],
    background: palette.lighter,
    borderColor: ({ color }) => (
      getColor(color)
    ),
    color: ({ color }) => (
      color === 'default' ? getColor(color, 'text') : getColor(color)
    ),
  };

  const inputElement = {
    textOverflow: 'ellipsis',
    width: '100%',
    border: 0,
    backgroundColor: 'transparent',
    height: 25,
    color: 'inherit',
    '&::placeholder': {
      color: () => (
        getColorLight(.75, 'default')
      ),
    },
    [active()]: {
      /**
       * @todo find an accessible solution
       */
      outline: 'none',
    },
  };

  const inputClear = {
    border: 0,
    lineHeight: 0,
    overflow: 'hidden',
    cursor: 'pointer',
    backgroundColor: 'transparent',
    [active()]: {
      /**
       * @todo find an accessible solution
       */
      outline: 'none',
      /**
       * @todo add support for transitions in ui-core
       */
      opacity: .5,
    },
  };

  const inputOrnament = {
    ...inlineFlexCenter,
    userSelect: 'none',
  };

  return {
    inputGroup,
    inputClear,
    inputOrnament,
    inputElement,
  };
};

const Input = (props) => {
  const {
    disabled,
    readonly,
    label,
    componentAtStart,
    componentAtEnd,
    gutter,
    clear,
    helper,
    error,
    errorMessage,
    type = 'text',
    color = 'default',
    onClear = () => {},
    onChange = () => {},
    value: defaultValue = '',
    ...factoryProps
  } = props;

  const [
    {
      inputGroup: classInputGroup,
      inputElement: classInputElement,
      inputClear: classInputClear,
      inputOrnament: classInputOrnament,
    },
] = useStyles(styles, {
    disabled,
    readonly,
    componentAtEnd,
    color: error ? 'error' : color,
  });

  let showClear = false;
  let showError = false;
  let showLabel = false;
  let showComponentAtStart = false;
  let showComponentAtEnd = false;

  if (clear) {
    showClear = true;
  }

  if (error) {
    showError = true;
    showClear = true;
  }

  if (label) {
    showLabel = true;
  }

  if (componentAtStart) {
    showComponentAtStart = true;
  }

  if (!clear && componentAtEnd) {
    showComponentAtEnd = true;
  }

  const [value, setValue] = useState(defaultValue);

  const inputRef = useRef();

  const handleChange = ({ currentTarget: { value } }) => {
    if (disabled) {
      return;
    }

    setValue(value);
    onChange({ value });
  }

  /**
   * @todo create event manager
   */
  const handleClear = () => {
    if (disabled) {
      return;
    }

    setValue('');
    onClear();
  }

  const handleControlClick = () => {
    inputRef.current.focus();
  };

  return (
    <Factory onClick={handleControlClick}>
      {showLabel && (
        <InputLabel children={label} />
      )}

      <Factory
        className={classInputGroup}
        marginBottom={gutter ?? 2}
        padding={1}
      >
        {showComponentAtStart && (
          <Factory
            element="span"
            className={classInputOrnament}
            paddingRight={1}
            children={componentAtStart}
          />
        )}

        <Factory
          ref={inputRef}
          element="input"
          className={classInputElement}
          type={type}
          disabled={disabled}
          readOnly={readonly}
          value={value}
          onChange={handleChange}
          {...factoryProps}
        />

        {showClear && (
          <Factory
            element="button"
            className={classInputClear}
            aria-label="Clear value"
            paddingX={1}
            onClick={handleClear}
          >
            <SvgIconClose
              color={error ? 'error' : color}
            />
          </Factory>
        )}

        {showComponentAtEnd && (
          <Factory
            element="span"
            className={classInputOrnament}
            paddingLeft={1}
            children={componentAtEnd}
          />
        )}
      </Factory>

      {showError && (
        <InputHelper
          color="error"
          children={errorMessage}
        />
      )}

      {helper && <InputHelper children={helper} />}
    </Factory>
  );
};

Input.propTypes = {
  clear: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool,
  ]),
  onClear: PropTypes.func,
  /**
   * @default text
   */
  type: PropTypes.oneOf([
    'text',
    'number',
    'phone',
    'email',
  ]),
  disabled: PropTypes.bool,
  readonly: PropTypes.bool,
  /**
   * @default Function
   */
  onChange: PropTypes.func,
  /**
   * @default default
   */
  color: PropTypes.oneOf([
    'default',
    'success',
    'warning',
    'error',
  ]),
  /**
   * @borrows Input.propTypes.label as InputLabel.propTypes.children
   */
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  componentAtStart: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
  ]),
  componentAtEnd: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
  ]),
  /**
   * @borrows Input.propTypes.helper as InputHelper.propTypes.children
   */
  helper: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  error: PropTypes.bool,
  errorMessage: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
};

export default Input;
