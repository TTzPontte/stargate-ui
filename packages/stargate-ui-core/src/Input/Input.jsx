import React, { useRef, useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { useStyles } from '@pontte/stargate-ui-styles';
import clsx from 'clsx';

import { Close as SvgIconClose } from '../icons';
import Factory from '../Factory';
import InputLabel from '../InputLabel';
import InputHelper from '../InputHelper';
import {
  zipCodeMask,
  currencyMask,
  phoneMask,
  ageMask,
  cpfMask,
  cnpjMask,
 } from './masks';

/**
 * @todo check color pattern for input element
 * @todo needs pattern for disabled and readonly state
 */
const styles = (theme) => {
  const {
    palette,
    radius,
    active,
    mode,
    breakpoints,
  } = theme;
  const { setLightness } = palette;

  const getColor = (color, type = 'color') => (
    palette?.[color]?.[mode][type]
  );

  const getColorLight = (n, color) => (
    setLightness(n, getColor(color))
  );

  const inlineFlexCenter = {
    display: 'inline-flex',
    alignItems: 'center',
  };

  const inputGroup = {
    ...inlineFlexCenter,
    overflow: 'hidden',
    borderRadius: radius(1.5),
    width: '100%',
    border: [[1, 'solid']],
    background: palette.lighter,
    borderColor: ({ color }) => (
      getColor(color)
    ),
    color: ({ color }) => (
      color === 'default' ? getColor(color, 'text') : getColor(color)
    ),
  };

  const inputElement = {
    textOverflow: 'ellipsis',
    width: '100%',
    border: 0,
    backgroundColor: 'transparent',
    height: 25,
    color: 'inherit',
    fontSize: 16,
    '&::placeholder': {
      color: () => (
        getColorLight(.75, 'default')
      ),
    },
    [active()]: {
      /**
       * @todo find an accessible solution
       */
      outline: 'none',
    },
    [breakpoints.down('md')]: {
      fontSize: 14,
    }
  };

  const inputClear = {
    border: 0,
    lineHeight: 0,
    overflow: 'hidden',
    cursor: 'pointer',
    backgroundColor: 'transparent',
    [active()]: {
      /**
       * @todo find an accessible solution
       */
      outline: 'none',
      /**
       * @todo add support for transitions in ui-core
       */
      opacity: .5,
    },
  };

  const inputOrnament = {
    ...inlineFlexCenter,
    userSelect: 'none',
  };

  return {
    inputGroup,
    inputClear,
    inputOrnament,
    inputElement,
  };
};

const Input = React.forwardRef((props, ref) => {
  const {
    disabled,
    readonly,
    label,
    componentAtStart,
    componentAtEnd,
    gutter,
    clear,
    helper,
    error,
    name,
    errorMessage,
    mask,
    type = 'text',
    color = 'default',
    onClear = () => {},
    onChange = () => {},
    onKeyUp= () => {},
    value: defaultValue = '',
    className: inheritedClassName,
    ...factoryProps
  } = props;

  const [
    {
      inputGroup: classInputGroup,
      inputElement: classInputElement,
      inputClear: classInputClear,
      inputOrnament: classInputOrnament,
    },
] = useStyles(styles, {
    disabled,
    readonly,
    componentAtEnd,
    color: error ? 'error' : color,
  });

  const className = clsx(classInputElement, inheritedClassName);

  let showClear = false;
  let showError = false;
  let showLabel = false;
  let showComponentAtStart = false;
  let showComponentAtEnd = false;

  if (clear) {
    showClear = true;
  }

  if (error) {
    showError = true;
    showClear = true;
  }

  if (label) {
    showLabel = true;
  }

  if (componentAtStart) {
    showComponentAtStart = true;
  }

  if (!clear && componentAtEnd) {
    showComponentAtEnd = true;
  }

  const [value, setValue] = useState(defaultValue);

  const inputRef = useRef(ref);

  const handleChange = (e) => {
    if (disabled) {
      return;
    }

    setValue(e.currentTarget.value);
    onChange(e);
  }

  const handleKeyUp = useCallback((e, mask) => {
    if (disabled) {
      return;
    }

    if (mask) {
      switch (mask) {
        case 'zipCode':
          zipCodeMask(e);
          break;
        case 'cpf':
          cpfMask(e);
          break;
        case 'phone':
          phoneMask(e);
          break;
        case 'age':
          ageMask(e);
            break;
        case 'currency':
          currencyMask(e);
          break;
        case 'cnpj':
          cnpjMask(e);
          break;
        default:
          console.log(`Sorry, we are out of ${mask}s.`);
      }
    }

  }, [disabled]);

  const handleClear = () => {
    if (disabled) {
      return;
    }

    setValue('');
    onClear();
  }

  const handleControlClick = () => {
    inputRef?.current !== null && inputRef?.current?.focus();
  };

  return (
    <Factory onClick={handleControlClick}>
      {showLabel && (
        <InputLabel children={label} />
      )}

      <Factory
        className={classInputGroup}
        marginBottom={gutter ?? 2}
        padding={1}
      >
        {showComponentAtStart && (
          <Factory
            element="span"
            className={classInputOrnament}
            paddingRight={1}
            children={componentAtStart}
          />
        )}

        <Factory
          ref={inputRef}
          element="input"
          className={className}
          type={type}
          disabled={disabled}
          readOnly={readonly}
          value={value}
          name={name}
          onChange={handleChange}
          onKeyUp={(e) => handleKeyUp(e, mask)}
          {...factoryProps}
        />

        {showClear && (
          <Factory
            element="button"
            className={classInputClear}
            aria-label="Clear value"
            paddingX={1}
            onClick={handleClear}
          >
            <SvgIconClose
              color={error ? 'error' : color}
            />
          </Factory>
        )}

        {showComponentAtEnd && (
          <Factory
            element="span"
            className={classInputOrnament}
            paddingLeft={1}
            children={componentAtEnd}
          />
        )}
      </Factory>

      {showError && (
        <InputHelper
          color="error"
          children={errorMessage}
        />
      )}

      {helper && <InputHelper children={helper} />}
    </Factory>
  );
});

Input.propTypes = {
  clear: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool,
  ]),
  onClear: PropTypes.func,
  /**
   * @default text
   */
  type: PropTypes.oneOf([
    'text',
    'number',
    'phone',
    'email',
  ]),
  disabled: PropTypes.bool,
  readonly: PropTypes.bool,
  /**
   * @default undefined
   */
  mask: PropTypes.string,
  /**
   * @default undefined
   */
  name: PropTypes.string.isRequired,
  /**
   * @default Function
   */
  onChange: PropTypes.func,
  /**
   * @default Function
   */
  onKeyUp: PropTypes.func,
  /**
   * @default default
   */
  color: PropTypes.oneOf([
    'default',
    'success',
    'warning',
    'error',
  ]),
  /**
   * @borrows Input.propTypes.label as InputLabel.propTypes.children
   */
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  componentAtStart: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
  ]),
  componentAtEnd: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
  ]),
  /**
   * @borrows Input.propTypes.helper as InputHelper.propTypes.children
   */
  helper: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  error: PropTypes.bool,
  errorMessage: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
};

export default Input;
