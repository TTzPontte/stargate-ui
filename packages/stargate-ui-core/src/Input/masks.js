export const zipCodeMask = (e) => {
  e.currentTarget.maxLength = 9;
  let value = e.currentTarget.value;
  value = value.replace(/\D/g, '');
  value = value.replace(/(\d{5})(\d)/, '$1-$2');
  e.currentTarget.value = value;
  return e;
};

export const currencyMask = (e) => {
  let value = e.currentTarget.value;
  value = value.replace(/\D/g, '');
  value = value.replace(/(\d)(\d{2})$/, '$1,$2');
  value = value.replace(/(?=(\d{3})+(\D))\B/g, '.');

  e.currentTarget.value = `R$ ${value}`;
  return e;
};

export const cpfMask = (e) => {
  let value = e.currentTarget.value;
  e.currentTarget.maxLength = 11;
  value = value.replace(/\D/g, '');
  value = value.replace(/(\d{3})(\d{3})(\d{3})(\d{2})/, '$1.$2.$3-$4');

  e.currentTarget.value = value;
  return e;
};

export const cnpjMask = (e) => {
  let value = e.currentTarget.value;
  e.currentTarget.maxLength = 14;
  value = value.replace(/\D/g, '');

  value = value.replace(
    /(\d{2})(\d{3})(\d{3})(\d{4})(\d{2})/,
    '$1.$2.$3/$4-$5'
  );

  e.currentTarget.value = value;
  return e;
};

export const ageMask = (e) => {
  let value = e.currentTarget.value;

  e.currentTarget.maxLength = 2;
  value = value.replace(/\D/g, '');

  e.currentTarget.value = value;
  return e;
};

export const phoneMask = (e) => {
  let value = e.currentTarget.value;

  e.currentTarget.maxLength = 11;
  value = value.replace(/\D/g, '');
  value = value.replace(/(\d{2})(\d{5})(\d{4})/, '($1) $2-$3');

  e.currentTarget.value = value;

  return e;
};

export const handleMask = (e, mask, setValue) => {
  if (mask) {
    switch (mask) {
      case 'zipCode':
        zipCodeMask(e);
        setValue(e.currentTarget.value);
        break;
      case 'cpf':
        cpfMask(e);
        setValue(e.currentTarget.value);
        break;
      case 'phone':
        phoneMask(e);
        setValue(e.currentTarget.value);
        break;
      case 'age':
        ageMask(e);
        setValue(e.currentTarget.value);
        break;
      case 'currency':
        currencyMask(e);
        setValue(e.currentTarget.value);
        break;
      case 'cnpj':
        cnpjMask(e);
        setValue(e.currentTarget.value);
        break;
      default:
        console.log(`Sorry, we are out of ${mask}s.`);
    }
  }
};
