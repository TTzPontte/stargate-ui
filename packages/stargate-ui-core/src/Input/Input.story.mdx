import {
  Meta,
  Canvas,
  Status,
  Story,
  Link,
  withAction,
} from '@pontte/stargate-ui-playground';
import {
  Typography,
  SvgIcon,
  Label,
  Close as SvgIconClose,
} from '@pontte/stargate-ui-core';

import Input from './Input';

export const InputStory = (props) => (
  <Input
    label="Cool input"
    onClear={withAction('onClear')}
    onChange={withAction('onChange')}
    {...props}
  />
);

<Meta title="Core/Form Elements/Input" component={Input} />

<Story
  name="Explore Properties"
  parameters={{
    docs: { disable: true },
    jest: ['Input.spec.jsx'],
  }}
>
  {InputStory.bind({})}
</Story>

## Input

<Status type="stable" />

### Basic

<Canvas>
  <Input
    label="Default"
    placeholder="Placeholder"
  />
  <Input
    label="Disabled"
    placeholder="Placeholder"
    disabled
    value="sahsauhsa"
  />
</Canvas>

### Using Icons And Ornaments

<Canvas>
  <Input
    label="Icon at start of input"
    componentAtStart={<SvgIconClose />}
  />
  <Input
    label="Icon at end of input"
    componentAtEnd={<SvgIconClose />}
  />
</Canvas>

### Custom Label

You can use any node element for `label` property. We recommend to use <Link kind="core-label--page">Label</Link> component.

<Canvas>
  <Input
    label={
      <Label>
        <SvgIconClose />
        Label customized
      </Label>
    }
  />
</Canvas>

### Clear Option

<Canvas>
  <Input
    label="Input with clear option"
    value="Mussum Ipsum, cacilds vidis litro abertis."
    clear
  />
</Canvas>

### Help Text

<Canvas>
  <Input
    label="Input with help text"
    helper="Mussum Ipsum, cacilds vidis litro abertis. Atirei o pau no gatis, per gatis num morreus. Interagi no mé, cursus quis, vehicula ac nisi. Mé faiz elementum girarzis, nisi eros vermeio. Nec orci ornare consequat. Praesent lacinia ultrices consectetur. Sed non ipsum felis."
  />
</Canvas>

### States And Messages For Form Validation

At the moment, **Input** only supports messages for error statement and 1 per time.

<Canvas>
  <Input
    label="Input With Error"
    value="Wrong value typed"
    error
    errorMessage="Mussum Ipsum, cacilds vidis litro abertis."
  />
</Canvas>

Stargate UI doesn't provide form validation control so your project needs to have it.

#### More States For Form Validation

<Canvas>
  <Input
    label="Input with success state"
    value="Mussum Ipsum, cacilds vidis litro abertis."
    color="success"
  />
  <Input
    label="Input with warning state"
    value="Mussum Ipsum, cacilds vidis litro abertis."
    color="warning"
  />
</Canvas>
