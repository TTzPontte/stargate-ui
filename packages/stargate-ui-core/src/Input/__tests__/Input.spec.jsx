import React from 'react';

import Input from '../Input';

describe('Input', () => {
  it('should render properly', () => {
    const { container } = render(
      <Input name="inputName" label="Lorem ipsum dolor" />
    );

    expect(container).toMatchSnapshot();
  });
});
