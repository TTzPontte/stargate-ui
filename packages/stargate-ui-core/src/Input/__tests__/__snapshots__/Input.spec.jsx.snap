// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render properly 1`] = `
<div>
  <div
    class="stargate-ui-factory-0-2-7 stargate-ui-factory-d0-0-2-9"
  >
    <label
      class="stargate-ui-factory-0-2-19 stargate-ui-factory-d0-0-2-21 stargate-ui-typography-0-2-12 stargate-ui-typography-d0-0-2-16 stargate-ui-label-0-2-11 stargate-ui-typographyFluid-0-2-15 stargate-ui-typographyFluid-d2-0-2-18"
    >
      Lorem ipsum dolor
    </label>
    <div
      class="stargate-ui-factory-0-2-23 stargate-ui-factory-d0-0-2-25 stargate-ui-inputGroup-0-2-1 stargate-ui-inputGroup-d0-0-2-5"
    >
      <input
        class="stargate-ui-factory-0-2-27 stargate-ui-factory-d0-0-2-29 stargate-ui-inputElement-0-2-4 stargate-ui-inputElement-d1-0-2-6"
        name="inputName"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
